lib/bricks-net-api-types.js

```diff
@@ -0,0 +1,1176 @@
+				/* istanbul ignore next: previous statement should throw */ break;
+		}
+		return _this;
+	}
+	
+	_this.has_body = false;
+	_this.body = "";
+	_this.content_type = "";
+	
+	cppArguments.assert('POST', [
+		[
+			cppArguments.assertion('string', 'const std::string&', 'url'),
+			function (url) {
+				HTTPRequestBase.call(_this, url);
+			}
+		],
+		[
+			cppArguments.assertion('string', 'const std::string&', 'url'),
+			cppArguments.assertion('string', 'const std::string&', 'body'),
+			cppArguments.assertion('string', 'const std::string&', 'content_type'),
+			function (url, body, content_type) {
+				HTTPRequestBase.call(_this, url);
+				
+				_this.has_body = true;
+				_this.body = body;
+				_this.content_type = content_type;
+			}
+		],
+		[
+			cppArguments.assertion('string', 'const std::string&', 'url'),
+			cppArguments.assertion('object', 'const T&', 'object'),
+			function (url, object) {
+				HTTPRequestBase.call(_this, url);
+				
+				_this.has_body = true;
+				_this.body = JSON.stringify({ data: object });
+				_this.content_type = "application/json";
+			}
+		]
+	], arguments);
+}
+inherits(POST, HTTPRequestBase);
+
+
+function POSTFromFile(url, file_name, content_type) {
+	var _this = this;
+	
+	if (!(_this instanceof POSTFromFile)) {
+		_this = new POSTFromFile(url, file_name, content_type);
+		return _this;
+	}
+	
+	_this.file_name = "";
+	_this.content_type = "";
+	
+	cppArguments.assert('POSTFromFile', [
+		[
+			cppArguments.assertion('string', 'const std::string&', 'url'),
+			cppArguments.assertion('string', 'const std::string&', 'file_name'),
+			cppArguments.assertion('string', 'const std::string&', 'content_type'),
+			function (url, file_name, content_type) {
+				HTTPRequestBase.call(_this, url);
+	
+				_this.file_name = file_name;
+				_this.content_type = content_type;
+			}
+		]
+	], arguments);
+}
+inherits(POSTFromFile, HTTPRequestBase);
+
+
+function HTTPResponse() {
+	var _this = this;
+	
+	// The final URL. Will be equal to the original URL, unless redirects have been allowed and took place.
+	_this.url = "";
+	// HTTP response code.
+	_this.code = 0;
+	
+	promiseHelpers.makeThenable(_this);
+	promiseHelpers.makeThenableProperty(_this, 'url');
+	promiseHelpers.makeThenableProperty(_this, 'code');
+}
+
+function HTTPResponseWithBuffer() {
+	HTTPResponse.call(this);
+	
+	this.body = "";
+	
+	promiseHelpers.makeThenableProperty(this, 'body');
+}
+inherits(HTTPResponseWithBuffer, HTTPResponse);
+
+function HTTPResponseWithResultingFileName() {
+	HTTPResponse.call(this);
+	
+	// The file name into which the returned HTTP body has been saved.
+	this.body_file_name = "";
+	
+	promiseHelpers.makeThenableProperty(this, 'body_file_name');
+}
+inherits(HTTPResponseWithResultingFileName, HTTPResponse);
+
+
+// Response storage policy.
+// The default one is `KeepResponseInMemory()`, which can be omitted.
+// The alternative one is `SaveResponseToFile(file_name)`.
+function KeepResponseInMemory() {
```

Comment:
I'd better not. I could abstract the logic of allowing to call a constructor without `new` (it's the `if (!(this instanceof ...)) { return new ...; }` part) after I've abstracted the logic of calling the constructor with an array of arguments.